Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,154 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Xml;
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security.Utilities;
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security
+{
+    public class NuGetCredentialsValidator : ValidatorBase
+    {
+        internal static NuGetCredentialsValidator Instance;
+
+        static NuGetCredentialsValidator()
+        {
+            Instance = new NuGetCredentialsValidator();
+        }
+
+        public static ValidationState IsValidStatic(ref string matchedPattern,
+                                                    ref Dictionary<string, string> groups,
+                                                    ref string failureLevel,
+                                                    ref string message,
+                                                    out Fingerprint fingerprint)
+        {
+            return ValidatorBase.IsValidStatic(Instance,
+                                               ref matchedPattern,
+                                               ref groups,
+                                               ref failureLevel,
+                                               ref message,
+                                               out fingerprint);
+        }
+
+        public static ValidationState IsValidDynamic(ref Fingerprint fingerprint, ref string message, ref Dictionary<string, string> options)
+        {
+            return IsValidDynamic(Instance,
+                                  ref fingerprint,
+                                  ref message,
+                                  ref options);
+        }
+
+        protected override ValidationState IsValidStaticHelper(ref string matchedPattern,
+                                                               ref Dictionary<string, string> groups,
+                                                               ref string failureLevel,
+                                                               ref string message,
+                                                               out Fingerprint fingerprint)
+        {
+            fingerprint = default;
+
+            if (!groups.TryGetNonEmptyValue("secret", out string secret) ||
+                !groups.TryGetNonEmptyValue("id", out string id) ||
+                !groups.TryGetNonEmptyValue("host", out string host))
+            {
+                return ValidationState.NoMatch;
+            }
+
+            if (StaticStringHelper.LikelyPowershellVariable(secret))
+            {
+                return ValidationState.NoMatch;
+            }
+
+            fingerprint = new Fingerprint
+            {
+                Host = host, // technically, this is XML which could contain multiple hosts.
```

Comment:
>Host [](start = 16, length = 4)

please us the order that others checks go, against, we waterfall from shortest name to longest, so if, host, secret, and platform.

Reply:
Done.

---
In reply to: [611238485](https://github.com/microsoft/sarif-pattern-matcher/pull/372#discussion_r611238485) [](ancestors = 611238485)